### DISCUSSION TITLE
Do executables in different containers share shared objects that are define in the same common image layer?

### DISCUSSION THREAD

**Discussion by GeoAzis** [containers/podman#23358] [Q&A]
I am investigating the memory impact of containerizing two processes that depends on the same shared object. I constructed the following experiment:

Process A depends on shared object glib.
Process B depends on shared object glib and slib.

## No containers used
First, I ran the processes on the host OS and I monitor the memory with free -m -s 2.
I observe that the free memory decreases by the size of the glib when I ran the A process and only by the size of slib when I ran the B process. Also, the cache increases by the same amounts at the same moment. If I interpret this correctly, the loaded shared object of process A is shared with Process B.

## Containers used
Then, I ran the processes in containers. Below you can see the setup:

Base container: 
```
# RHEL8 Universal Base Image as a starting point
FROM ubi8
COPY libglib.so /app/
COPY libslib.so /app/
```

Process A container:
```
FROM base
# Precompiled executable
COPY dependsong /executable/
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/app
```

Process B container:
```
FROM localhost/base
# Precompiled executable
COPY dependsongands /executable/
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/app
```

When I run the processes inside the containers and monitor the memory (again free -m -s 2 from my host OS), I see quite different results. It appears that glib is loaded twice (I see an equivalent decrease in the free memory twice and an equivalent increase in cache). 

According to the documentation I found online, the sharing of shared objects depends on the storage driver you are using. I am using the overlay storage driver, which suggests that sharing should occur. However, that's appears not to be the case. 

Below you can find the podman info output related to the graphDriver:
```
graphDriverName: overlay
graphOptions:
   overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.11-1.module+el8.8.0+21336+3fe03132.x86_64
      Version: |-
         fusermount3 version: 3.3.0
         fuse-overlayfs: version 1.11
         FUSE library version 3.3.0
         using FUSE kernel interface version 7.26
graphStatus:
   Backing Filesystem: xfs
   Native Overlay Diff: "false"
   Supports d_type: "true"
   Using metacopy: "false"
```

I have the following questions:

1. Am I misinterpreting the results?
2. Does the described behavior match your expectations?
3. What could I try to ensure sharing occurs?

**Comment by rhatdan** [comment]
@giuseppe PTAL
